pkg/loop/internal/broker.go

```diff
@@ -106,15 +108,19 @@ func (b *brokerExt) serve(name string, register func(*grpc.Server), deps ...reso
 		server = b.NewServer(nil)
 	}
 	register(server)
+	b.wg.Add(1)
```

Comment:
Sharing the `WaitGroup` between method calls seems problematic (what prevents `Add` from being called after `Wait`?). Can we just use separate instances instead?
```suggestion
	var wg sync.WaitGroup
	b.wg.Add(1)
```

Reply:
We could use separate instances, but I think the execution order of `serve` guarantees that `Add` is not being called after `Wait`. `Add` is called when each gr is created (but in the same go routine as `serve`), and `Wait` is only called once `serve` returns. 

I can remove it as a field of `brokerExt` though - it's only essential to `serve`.

Reply:
> We could use separate instances, but I think the execution order of serve guarantees that Add is not being called after Wait. Add is called when each gr is created (but in the same go routine as serve), and Wait is only called once serve returns.

But `serve` can be called multiple times. Isn't it possible for an early `serve` call to close before a later `serve` call is made?

Reply:
I've refactored it out of brokerExt, so I think this should be taken care of?

Reply:
> But serve can be called multiple times. Isn't it possible for an early serve call to close before a later serve call is made?

As long as serve isn't called concurrently! I think you can reuse a waitgroup once `Wait()` is called and the internal counter hits 0.

```
package main

import "sync"
import "fmt"

type Example struct {
	wg sync.WaitGroup
}

func main() {
	ex := &Example{}

	ex.callTwice()
	ex.callTwice()

}

func (e *Example) callTwice() {
	e.wg.Add(1)
	go func() {
		fmt.Println("Hi")
		defer e.wg.Done()
	}()

	e.wg.Wait()
}
```

Reply:
It can be used concurrently and I thought our tests exercised that and would trigger a race. Perhaps with more load.